enclave/impl/src/message_bus/mod.rs

```diff
@@ -0,0 +1,753 @@
+                challenge[0] = !challenge[0];
+
+                assert!(bus.handle_query(&req, &challenge).is_err());
+            }
+        });
+    }
+
+    #[test_with_logger]
+    fn test_query_bounds_checks(logger: Logger) {
+        let mut bus = get_test_bus(logger);
+
+        run_with_several_seeds(|mut rng| {
+            let secret = RistrettoPrivate::from_random(&mut rng);
+
+            // Note: deletes will "fail" when the msg id doesn't exist, but they fail in constant-time and not with a "hard" error.
+            for request_type in &[
+                REQUEST_TYPE_CREATE,
+                REQUEST_TYPE_READ,
+                REQUEST_TYPE_UPDATE,
+                REQUEST_TYPE_DELETE,
+            ] {
+                let (challenge, mut req) = sign_query(
+                    &secret,
+                    *request_type,
+                    FromRandom::from_random(&mut rng),
+                    &mut rng,
+                );
+
+                // Properly signed query should work
+                bus.handle_query(&req, &challenge).unwrap();
+
+                // Message id which is too long should be an error
+                req.record.msg_id.resize(17, 0u8);
+                assert!(bus.handle_query(&req, &challenge).is_err());
+
+                // Message id which is too short should be an error
+                req.record.msg_id.resize(15, 0u8);
+                assert!(bus.handle_query(&req, &challenge).is_err());
+
+                // Recipient which is too long should be an error
+                req.record.msg_id.resize(16, 0u8);
+                req.record.recipient.resize(33, 0u8);
+                assert!(bus.handle_query(&req, &challenge).is_err());
+
+                // Recipient which is too short should be an error
+                req.record.recipient.resize(31, 0u8);
+                assert!(bus.handle_query(&req, &challenge).is_err());
+
+                // Payload which is too long should be an error
+                req.record.recipient.resize(32, 0u8);
+                req.record.payload.resize(937, 0u8);
+                assert!(bus.handle_query(&req, &challenge).is_err());
+
+                // Payload which is too short should be an error
+                req.record.payload.resize(935, 0u8);
+                assert!(bus.handle_query(&req, &challenge).is_err());
+
+                // Request type of zero is an error
+                req.record.payload.resize(936, 0u8);
+                req.request_type = 0;
+                assert!(bus.handle_query(&req, &challenge).is_err());
+            }
+        });
+    }
+
+    #[test_with_logger]
+    fn create_and_read_visibility(logger: Logger) {
+        let mut bus = get_test_bus(logger);
+
+        run_with_several_seeds(|mut rng| {
+            let secret1 = RistrettoPrivate::from_random(&mut rng);
+            let secret2 = RistrettoPrivate::from_random(&mut rng);
+            let secret3 = RistrettoPrivate::from_random(&mut rng);
```

Comment:
I think it will be a bit easier to understand if these are renamed to identity1/2/3